src/backend/plaf.rs

```diff
@@ -0,0 +1,507 @@
+
+#[derive(Clone, Debug)]
+pub struct ChiquitoPlaf<F: PrimeField, TraceArgs, StepArgs: Clone> {
+    debug: bool,
+    circuit: cCircuit<F, TraceArgs, StepArgs>,
+    // Chiquito column id doesn't start from zero.
+    // Plaf column index starts from 0 for each column type (advice, fixed, and instance).
+    // Therefore a mapping is needed to convert chiquito column id to plaf index.
+    c_column_id_to_p_column_index: HashMap<u32, usize>,
+    advice_index: usize, // This is Plaf's index of witness (advice) column.
+    fixed_index: usize,  // This is Plaf's index of fixed column.
+}
+
+impl<F: PrimeField<Repr = [u8; 32]>, TraceArgs, StepArgs: Clone>
+    ChiquitoPlaf<F, TraceArgs, StepArgs>
+{
+    // <Repr = [u8; 32]> is required by `from` function in the following line:
+    // `cPolyExpr::Halo2Expr(e) => pExpr::from(e)`.
+    // This function converts a halo2 Expression<F> to a polyexen `Expr<PlonkVar>`.
+    // Therefore, `F: PrimeField<Repr = [u8; 32]>` is required.
+    pub fn new(
+        circuit: cCircuit<F, TraceArgs, StepArgs>,
+        debug: bool,
+    ) -> ChiquitoPlaf<F, TraceArgs, StepArgs> {
+        ChiquitoPlaf {
+            debug,
+            circuit,
+            c_column_id_to_p_column_index: HashMap::new(),
+            advice_index: 0,
+            fixed_index: 0,
+        }
+    }
+
+    pub fn get_plaf(&mut self, k: u32) -> Plaf {
+        let mut plaf = Plaf::default();
+        let p = get_field_p::<F>();
+        plaf.info.p = p;
+
+        plaf.info.num_rows = 2usize.pow(k);
+
+        let mut c_column_id_to_p_column_index = HashMap::<u32, usize>::new();
+        let mut advice_index = 0;
+        let mut fixed_index = 0;
+
+        for column in self.circuit.columns.iter() {
+            if self.debug {
+                println!("annotation: {}, id: {}", column.annotation, column.id);
+            }
+            self.convert_and_push_plaf_column(
+                column,
+                &mut plaf,
+                Some(&mut c_column_id_to_p_column_index),
+                Some(&mut advice_index),
+                Some(&mut fixed_index),
+            );
+            if self.debug {
+                println!("MAP {:#?}", c_column_id_to_p_column_index);
+            }
+        }
+
+        self.c_column_id_to_p_column_index = c_column_id_to_p_column_index;
+        self.advice_index = advice_index;
+        self.fixed_index = fixed_index;
+
+        if !self.circuit.polys.is_empty() {
+            for c_poly in &mut self.circuit.polys.iter() {
+                let plaf_poly = pPoly {
+                    name: c_poly.annotation.clone(),
+                    exp: self.convert_plaf_poly(&c_poly.expr),
+                };
+                plaf.polys.push(plaf_poly);
+            }
+        }
+
+        for lookup in self.circuit.lookups.iter() {
+            let v1 = lookup
+                .exprs
+                .clone()
+                .into_iter()
+                .map(|(e1, _)| self.convert_plaf_poly(&e1))
```

Comment:
.map(|(e1, e2)| (self.convert_plaf_poly(&e1), self.convert_plaf_poly(&e2)))

Reply:
Better names for v1, v2

Reply:
Fixed. As discussed, the above won't work as it's converting a vector of tuples to a tuple of vectors. Used `fold` instead.